test/helper/common.bash

```diff
@@ -172,40 +172,38 @@ function wait_for_service() {
     container_has_service_running "${CONTAINER_NAME}" "${SERVICE_NAME}"
 }
 
-function wait_for_changes_to_be_detected_in_container() {
-  local CONTAINER_NAME="${1}"
-  local TIMEOUT=${TEST_TIMEOUT_IN_SECONDS}
+# NOTE: Relies on ENV `LOG_LEVEL=debug` or higher
+function _wait_until_expected_count_is_matched() {
+  function __get_count() {
+    docker exec "${CONTAINER_NAME}" grep --count "${MATCH_CONTENT}" "${MATCH_IN_LOG}"
+  }
 
-  # shellcheck disable=SC2016
-  repeat_in_container_until_success_or_timeout "${TIMEOUT}" "${CONTAINER_NAME}" bash -c 'source /usr/local/bin/helpers/index.sh; _obtain_hostname_and_domainname; cmp --silent -- <(_monitored_files_checksums) "${CHKSUM_FILE}" >/dev/null'
-}
+  # WARNING: Keep in mind it is a '>=' comparison.
+  # If you provide an explict count to match, ensure it is not too low to cause a false-positive.
+  function __has_expected_count() {
+    [[ $(__get_count) -ge "${EXPECTED_COUNT}" ]]
+  }
+
+  local CONTAINER_NAME=$1
+  local EXPECTED_COUNT=$2
 
-# NOTE: Relies on ENV `LOG_LEVEL=debug` or higher
-function wait_until_change_detection_event_completes() {
-  local CONTAINER_NAME="${1}"
   # Ensure early failure if arg is missing:
-  assert_not_equal "${CONTAINER_NAME}" ""
+  assert_not_equal "${CONTAINER_NAME}" ''
 
   # Ensure the container is configured with the required `LOG_LEVEL` ENV:
   assert_regex \
     $(docker exec "${CONTAINER_NAME}" env | grep '^LOG_LEVEL=') \
     '=(debug|trace)$'
 
-  # NOTE: Change events can start and finish all within < 1 sec,
-  # Reliably track the completion of a change event by comparing the before/after count:
-  function __change_event_count() {
-    docker exec "${CONTAINER_NAME}" grep --count "${CHANGE_EVENT_END}" /var/log/supervisor/changedetector.log
-  }
-
-  function __is_changedetector_finished() {
-    [[ $(__change_event_count) -gt "${NUM_CHANGE_EVENTS_BEFORE}" ]]
-  }
-
-  # Count by completions of this debug log line from `check-for-changes.sh`:
-  local CHANGE_EVENT_END='Completed handling of detected change'
-  local NUM_CHANGE_EVENTS_BEFORE=$(__change_event_count)
+  # Default behaviour is to wait until one new match is found (eg: incremented),
+  # unless explicitly set (useful for waiting on a min count to be reached):
+  if [[ -z $EXPECTED_COUNT ]]
+  then
+    # +1 of starting count:
+    EXPECTED_COUNT=$(bc <<< "$(__get_count) + 1")
+  fi
 
-  repeat_until_success_or_timeout 60 __is_changedetector_finished
+  repeat_until_success_or_timeout 20 __has_expected_count
```

Comment:
I haven't compared, but [casper had suggested a different approach than polling](https://github.com/docker-mailserver/docker-mailserver/pull/2988#discussion_r1064246820) like I'm doing here.

```suggestion
  timeout 60 docker exec "${CONTAINER_NAME}" bash -c "tail -F '${MATCH_IN_LOG}' | grep --max-count 1 '${MATCH_CONTENT}'"
```

Reply:
I cannot reproduce the hanging. I tried this:

```bash
# exit as soon as the keyword "spawned" appears
docker logs -n0 -f mail | grep -q spawned
```

Then in another session:

`docker exec mail supervisorctl restart update-check`

Reply:
I will give it a look again, perhaps there was something else affecting it. When I tried it was with an Amavis test unrelated to change detection, so I assumed it'd be the same issue if used elsewhere. I'll see if I can reproduce again :+1:

Reply:
I was able to reproduce on current master the same issue. I tested with replacing this:

https://github.com/docker-mailserver/docker-mailserver/blob/1650cdf76f0b1067192d0d12f92508b561a87597/test/tests/parallel/set1/spam_virus/spam_junk_folder.bats#L33-L35

```bash
# Passes before timing out:
timeout 20 docker exec "${CONTAINER_NAME}" bash -c "tail -F /var/log/mail/mail.log | grep --max-count 1 'Passed SPAM'"
# Same result here:
docker exec "${CONTAINER_NAME}" bash -c "timeout 20 tail -F /var/log/mail/mail.log | grep --max-count 1 'Passed SPAM'"
# Times out with failure, and due to -q has no output:
timeout 20 bash -c "docker logs --follow ${CONTAINER_NAME} | grep -q 'Passed SPAM'"
# Times out with failure, but correctly shows output of requested max lines (even if there were more present)
timeout 20 bash -c "docker logs --follow ${CONTAINER_NAME} | grep --max-count 1 'Passed SPAM'"
# Times out but **passes** (docker logs was stopped by timeout, grep was successful)
timeout 20 docker logs --follow "${CONTAINER_NAME}" | grep --max-count 1 'Passed SPAM'
```

```bash
function _passed_spam() {
  # Any method from above here
}

# Replace as 2nd assert in either test case of `tests/parallel/set1/spam_virus/spam_junk_folder.bats:
run _passed_spam
assert_success
```

---

Perhaps it's something different with our environments or how `docker logs --follow`  works on our machines? For me, given the above observations, it doesn't seem like the stream is closed?

I had also verified by running the same command in another terminal session directly while the test was running. It would find the match and output it, but not exit until the timeout was reached.

I've had a similar experience with `-q` (not as nice as `--max-count`), where either option was working at least for `decoder` as the value to match when the container was starting up (_I could not reproduce with `Passed SPAM` for some reason.._). I could repeat the command in the terminal several times and it'll exit quickly with either option successful. However a few seconds after that and it would fail, even though the logs command would clearly show the lines :shrug: 

Clearly something is iffy there, at least on my end :\

Reply:
I swapped `grep` for [`rg` (ripgrep)](https://github.com/BurntSushi/ripgrep), and now it works as you'd expect...

So something wrong with my `grep` command I guess? :man_shrugging: 

---

I'd rather avoid the `-q` or `--max-count` approach, at least with a local grep call, as it seems that's where the problem is. The containers grep was working fine with `tail`, which would be ok as it avoids false positives when I run tests locally.

We could add `ripgrep` as a dependency for testing like we do with `jq` I suppose? (_but then CI may need to grab/install it each time, unless it already provides that like it does `jq`_)

Reply:
> So something wrong with my grep command I guess? 🤷‍♂️

yes 😄 

`docker logs ${CONTAINER_NAME} | grep 'Passed SPAM {RelayedTaggedInbound,Quarantined}'"`

When you don't pass `-q` or `--max-count 1`, grep runs infinitly. Why should it stop?

If you only need to now, if a keyword was found, go with `-q`. If you need the output to parse further or make assertions, go with `--max-count 1`.

Reply:
> When you don't pass `-q` or `--max-count 1`, grep runs infinitly. Why should it stop?

You're referring to the current version in master I referenced? That is not using `--follow`, so it does not run infinitely. The attempts I tried can be found in the collapsed "> Reproduction" section.

I collapsed it after responding with an update about my local `grep` being the problem. I understand the reason for `-q` or `--max-count 1`, but only if we're using `grep` within the container (_doesn't make sense for `docker logs --follow` due to that_), as for whatever reason my local `grep` is misbehaving which would be problematic for me running tests locally :sweat_smile: (_something I've been doing quite frequently lately_)

If we're to use either option with `docker logs --follow`, then I'd need to know why my `grep` is broken and how to fix it, or add `ripgrep` as a dependency for tests.

Reply:
> The attempts I tried can be found in the collapsed "> Reproduction" section.

Thanks, I overlooked that.

BTW: Are you running the tests on debian/ubuntu?

Reply:
> Are you running the tests on debian/ubuntu?

EndeavourOS (ArchLinux based)

Reply:
Depending on how much more time you want to invest, you may try the "grep" binary from debian:

```
docker create --name grep debian
docker cp grep:/bin/grep .
docker rm grep
```

I am wondering, that such a basic tool behaves differently between linux distros.

PS: If we are ever in doubt, we should stick with debian based distros supported for testing.